### DISCUSSION TITLE
FastAPI default configuration leaks traffic information to multiple third parties

### DISCUSSION THREAD

**Discussion by magic-david** [fastapi/fastapi#8980] [Questions]
### First check

* [x] I added a very descriptive title to this issue.
* [ ] I used the GitHub search to find a similar issue and didn't find it.
* [ ] I searched the FastAPI documentation, with the integrated search.
* [ ] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [ ] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [ ] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [ ] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

```Python
from fastapi import FastAPI

app = FastAPI()
```

### Description

- Visiting `/docs` or `/redocs` in the example application leaks browser fingerprint, IP address and referer headers to multiple third party sites.

- Using the documentation offline is impossible by default.

### Environment

* FastAPI 0.62.0

* Python 3.8

### Additional context

The FastAPI integrated documentation, which is pretty much its headline feature, by default depends on:

- a favicon loaded from tiangolo.com
- Javascript libraries loaded from jsdelivr.net

This leaks internal hostnames to two third party sites, in addition to creating an artificial dependency on working DNS, SSL and networking to any FastAPI application. Per https://github.com/tiangolo/fastapi/issues/382, although it is possible to self-host, the procedure involved is significantly less than convenient, sufficiently so that an alternative solution may be preferred over FastAPI.

There are two major problems with this default that would induce me to vocally discourage further use of the framework internally:

- Security-wise it is a huge red flag, and makes it factual to say that FastAPI is not secure by default. One must wonder what else there is to discover about FastAPI or its choice of dependencies, that the average user has no time to discover.

- The favicon in particular amounts to an undocumented web stats tracker embedded in every FastAPI application. It does not matter whether the traffic logs on the server side are treated in this manner, the existence of the tracker is completely unacceptable. Accidental conveniences like this are far too common on the web to be treated as unintentional.

I would be willing to submit a PR to reduce https://fastapi.tiangolo.com/advanced/extending-openapi/#self-hosting-javascript-and-css-for-docs to a one-liner, but would be even happier to produce a PR making it the default behaviour. Since the latter case would require a new dependency on `aiofiles`, perhaps it is worth discussing first if there is any interest from the project.

**Comment by ycd** [comment]
If anyone is willing to change this, it can be changed by this code piece.


```python
from fastapi import FastAPI
from fastapi.openapi.docs import get_swagger_ui_html, get_redoc_html


app = FastAPI(docs_url=None, redoc_url=None)


@app.get("/docs", include_in_schema=False)
def overridden_swagger():
	return get_swagger_ui_html(openapi_url="/openapi.json", title="FastAPI", swagger_favicon_url="https://placekitten.com/200/300")

@app.get("/redoc", include_in_schema=False)
def overridden_redoc():
	return get_redoc_html(openapi_url="/openapi.json", title="FastAPI", redoc_favicon_url="https://placekitten.com/200/300")
```

**Comment by actualben** [comment]
I agree that non-tracking should be the default (somehow).

For others coming to this issue, <https://fastapi.tiangolo.com/advanced/extending-openapi/#self-hosting-javascript-and-css-for-docs> is almost the whole solution, but it is missing a few things:

* `swagger_favicon_url` on the call to `get_swagger_ui_html`
* `redoc_favicon_url` on the call to `get_redoc_html`
* `with_google_fonts=False` on the call to `get_redoc_html`

Here's a version with those three items included:

```python
from fastapi import FastAPI
from fastapi.openapi.docs import (
    get_redoc_html,
    get_swagger_ui_html,
    get_swagger_ui_oauth2_redirect_html,
)
from fastapi.staticfiles import StaticFiles

app = FastAPI(docs_url=None, redoc_url=None)

app.mount("/static", StaticFiles(directory="static"), name="static")


@app.get("/docs", include_in_schema=False)
async def custom_swagger_ui_html():
    return get_swagger_ui_html(
        openapi_url=app.openapi_url,
        title=app.title + " - Swagger UI",
        oauth2_redirect_url=app.swagger_ui_oauth2_redirect_url,
        swagger_js_url="/static/swagger-ui-bundle.js",
        swagger_css_url="/static/swagger-ui.css",
        swagger_favicon_url="/favicon.ico",  # or /static/favicon.ico if applicable
    )


@app.get(app.swagger_ui_oauth2_redirect_url, include_in_schema=False)
async def swagger_ui_redirect():
    return get_swagger_ui_oauth2_redirect_html()


@app.get("/redoc", include_in_schema=False)
async def redoc_html():
    return get_redoc_html(
        openapi_url=app.openapi_url,
        title=app.title + " - ReDoc",
        redoc_js_url="/static/redoc.standalone.js",
        redoc_favicon_url="/favicon.ico",  # or /static/favicon.ico if applicable
        with_google_fonts=False,
    )
```

You can get the static assets needed with this curl command (from the static directory):

```sh
curl \
  --compressed \
  --remote-name-all \
  --tlsv1.2 \
  "https://cdn.jsdelivr.net/npm/swagger-ui-dist@3/swagger-ui-bundle.js" \
  "https://cdn.jsdelivr.net/npm/swagger-ui-dist@3/swagger-ui.css" \
  "https://cdn.jsdelivr.net/npm/swagger-ui-dist@3/swagger-ui.css.map" \
  "https://cdn.jsdelivr.net/npm/redoc@next/bundles/redoc.standalone.js"
```

... but I wonder if it makes more sense to manage them with npm (or something else with a lockfile).
